plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryClient.java

```diff
@@ -82,20 +86,43 @@
     private final Optional<String> configProjectId;
 
     public BigQueryClient(
-            BigQuery bigQuery,
+            HeaderProvider headerProvider,
+            Optional<Credentials> credentials,
             boolean caseInsensitiveNameMatching,
             ViewMaterializationCache materializationCache,
             Duration metadataCacheTtl,
+            Optional<String> configParentProjectId,
             Optional<String> configProjectId)
     {
-        this.bigQuery = requireNonNull(bigQuery, "bigQuery is null");
         this.materializationCache = requireNonNull(materializationCache, "materializationCache is null");
         this.caseInsensitiveNameMatching = caseInsensitiveNameMatching;
         this.remoteDatasetCache = EvictableCacheBuilder.newBuilder()
                 .expireAfterWrite(metadataCacheTtl.toMillis(), MILLISECONDS)
                 .shareNothingWhenDisabled()
                 .build(CacheLoader.from(this::listDatasetsFromBigQuery));
         this.configProjectId = requireNonNull(configProjectId, "projectId is null");
+        String parentProjectId = calculateParentProjectId(configParentProjectId, credentials);
+
+        BigQueryOptions.Builder options = BigQueryOptions.newBuilder()
+                .setHeaderProvider(headerProvider)
+                .setProjectId(parentProjectId);
+        credentials.ifPresent(options::setCredentials);
+        this.bigQuery = options.build().getService();
+    }
+
+    // Note that at this point the config has been validated, which means that option 2 or option 3 will always be valid
+    static String calculateParentProjectId(Optional<String> configParentProjectId, Optional<Credentials> credentials)
+    {
+        // 1. Get from configuration
+        return configParentProjectId
+                // 2. Get from the provided credentials, but only ServiceAccountCredentials contains the project id.
+                // All other credentials types (User, AppEngine, GCE, CloudShell, etc.) take it from the environment
+                .orElseGet(() -> credentials
+                        .filter(ServiceAccountCredentials.class::isInstance)
+                        .map(ServiceAccountCredentials.class::cast)
+                        .map(ServiceAccountCredentials::getProjectId)
+                        // 3. No configuration was provided, so get the default from the environment
+                        .orElseGet(BigQueryOptions::getDefaultProjectId));
```

Comment:
It's a little hard to understand the actual change for a bug fix because this commit moves some methods. Could you move back the method or separate a commit?

Reply:
Yeah I'll split them; Can this one still make it to 408?

Reply:
[a6c1c23](https://github.com/trinodb/trino/pull/15868/commits/a6c1c231041a82863552729fd56d23c988303fb7) Relocating the BigQuery client logic under `BigQuery` wrapper makes a bit more sense to me, but if you prefer, we can scrap that and possibly move it to a separate PR.

Reply:
It would be if you can separate a PR.

Reply:
Fair enough. Done @ebyhr lmk if you need anything else.